src/Data/Array/Mutable/Linear.hs

```diff
@@ -60,31 +64,42 @@ import GHC.Exts hiding (toList, fromList)
 import GHC.Stack
 import qualified Unsafe.Linear as Unsafe
 import qualified Unsafe.MutableArray as Unsafe
-import Prelude.Linear ((&))
-import Prelude hiding (length, read)
+import Prelude.Linear ((&), forget)
+import qualified Control.Monad as Control
+import Prelude hiding (size, read)
 import qualified Prelude
 
 -- # Data types
 -------------------------------------------------------------------------------
 
 data Array a where
-  Array :: Int -> (MutableArray# RealWorld a) -> Array a
+  Array :: MutableArray# RealWorld a -> Array a
 
 -- # Creation
 -------------------------------------------------------------------------------
 
+-- | Creates an empty array.
+empty :: (Array a #-> Unrestricted b) -> Unrestricted b
+empty f = f (Array (Unsafe.emptyMutArray ()))
+
 -- | Allocate a constant array given a size and an initial value
--- The size must be greater than zero, otherwise this errors.
-alloc :: HasCallStack =>
-  Int -> a -> (Array a #-> Unrestricted b) -> Unrestricted b
-alloc size x f
-  | size > 0 = f (Array size (Unsafe.newMutArr size x))
+-- The size must be non-negative, otherwise this errors.
+alloc :: Int -> a -> (Array a #-> Unrestricted b) -> Unrestricted b
+alloc size val f
+  | size >= 0 = f (Array (Unsafe.newMutArr size val))
   | otherwise = error $ "Trying to allocate an array of size " ++ show size
 
--- | Allocate an array from a non-empty list (and error on empty lists)
+-- | Allocate a constant array given a size and an initial value,
+-- using another array as a uniqueness proof.
+allocBeside :: Int -> a -> Array b #-> (Array b, Array a)
+allocBeside size val orig
+  | size >= 0 = (orig, Array (Unsafe.newMutArr size val))
+  | otherwise = orig `lseq` error ("Trying to allocate an array of size " ++ show size)
```

Comment:
I find this function interesting. It is required since when implementing HashMap's, insertions can cause a new, bigger array to be created. However, we can not use `alloc` function there since it requires a continuation; and making HashMap insertions take a continuation would be unwieldy. 

I wanted to replace it with something more generic, but couldn't figure out how to do it nicely. I think this connects to @aspiwack 's suggestion on https://github.com/tweag/linear-base/issues/130.